### PR TITLE
Added PermissibleBase getters and setters

### PR DESCRIPTION
**Pull request by Badbird5907**
Adding this API will remove the need to use reflections in plugins to get and set `PermissibleBase` on `CraftHumanEntity` to override permissions logic

### CODE REVIEW

**Comment by Owen1212055** [comment]
Note: https://github.com/PaperMC/Paper/pull/8108

Do note that this expands the whole permission overriding and allows you to override the permission logic in the SimplePluginManager.

**Comment by Badbird5907** [comment]
Added the missing annotations causing tests to fail

File: patches/api/0400-Added-PermissibleBase-getter-and-setter.patch [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index 1d98abff1ad0116f7a2599f078aa730cb84843c1..d0dec0999a2ffe0117b88a9bddf3b707bcd16710 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -33,6 +33,7 @@ import org.bukkit.event.player.PlayerResourcePackStatusEvent;
+ import org.bukkit.inventory.EquipmentSlot;
+ import org.bukkit.inventory.ItemStack;
+ import org.bukkit.map.MapView;
++import org.bukkit.permissions.PermissibleBase;
```
Comment (kennytv) [inline patches/api/0400-Added-PermissibleBase-getter-and-setter.patch]: Inline the import

File: patches/api/0400-Added-PermissibleBase-getter-and-setter.patch [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ import org.bukkit.plugin.Plugin;
+ import org.bukkit.plugin.messaging.PluginMessageRecipient;
+ import org.bukkit.scoreboard.Scoreboard;
+@@ -1965,7 +1966,7 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+      * @param saturationLevel the saturation level of the player
+      */
+     public void sendHealthUpdate(final double health, final int foodLevel, final float saturationLevel);
+-
```
Comment (kennytv) [inline patches/api/0400-Added-PermissibleBase-getter-and-setter.patch]: Don't change indents/spacing

**Review by kennytv — CHANGES_REQUESTED** [review]

**Comment by Badbird5907** [comment]
Did the requested changes
_idea didn't like the indents_

File: patches/server/0926-Added-PermissibleBase-getter-and-setter.patch [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+index e779dcc4982ff51e4d450265fd61bc26e8e74d3a..61fd80c691476c3eaec9ff182715913da3a0f051 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+@@ -70,7 +70,7 @@ import org.bukkit.plugin.Plugin;
+ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
+     private CraftInventoryPlayer inventory;
+     private final CraftInventory enderChest;
+-    protected final PermissibleBase perm = new PermissibleBase(this);
++    protected PermissibleBase perm = new PermissibleBase(this);
```
Comment (Lulu13022002) [inline patches/server/0926-Added-PermissibleBase-getter-and-setter.patch]: Missing the paper comment

File: patches/api/0400-Added-PermissibleBase-getter-and-setter.patch [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+@@ -2832,4 +2832,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     @Override
+     Spigot spigot();
+     // Spigot end
++
++    // Paper start
++    /**
++     * Set the player's {@link org.bukkit.permissions.PermissibleBase}
++     * @param base {@link org.bukkit.permissions.PermissibleBase}
++     * @return the old {@link org.bukkit.permissions.PermissibleBase}
```
Comment (Lulu13022002) [inline patches/api/0400-Added-PermissibleBase-getter-and-setter.patch]: These links seems a little bit too redundant, maybe keep only the links into the description but not into the parameter explanation ?

**Comment by zml2008** [comment]
this feels like the kind of thing that exposes internal details of api impl which makes evolving the permissions api harder, and makes it easier for plugin developers to make mistakes without any traceability.

you'd be better off reworking how permissions are set and computed in the first place to remove the numerous footguns developers encounter when customizing permissions api logic - things like state tracking across the various join/quit/kick events, the permissions impact of recalculating permissions, and all the logic with default permissions

**Comment by electronicboy** [comment]
The thing with evolving the perm API is that it's generally one of the many things which has been left to rot, hence why any perm plugin of note will literally just rip the thing out in order to add basic features, such as verbose mode, wildcards, etc

I'm much open towards expanding this API, but, I'd much rather that be an effort kicked off through perm plugin devs whom know their needs here so that we can actually deliver towards it, but, at this point, people reflect into this mess anyways that I don't have a strong urge to deny this personally, but, did wanna ask people like Luck if they had any thoughts or considerations; It's not like exposing this in the API is a huge priority for us and I'd much rather see perm plugin devs discuss this sorta thing

**Comment by Owen1212055** [comment]
Closing, as in general as specified above it's best to instead expand this API instead of trying to move around it like this.

**Pull request closed without merging**